### ISSUE TITLE
Request: initially suggest "standard user specific" folder for download

### ISSUE THREAD

**Issue by porres** [pure-data/deken#156]
Howdy, if I understand it correctly, deken searches for "standard paths" and prompts you for the first one it finds. If it can't find a valid one, it suggests the creation of the "standard user specific" folder. This seems to be working since this commit: https://github.com/pure-data/deken/commit/8625a443917f25448b5b1199b0507f5fe8c59607

This is extremely important for windows, as there was no valid download path deken could find and now this is solved! I understand there was a concern on automatically adding "standard folders", but this solution did meet halfway, and allowed the creation of such folder at request. 

But unfortunately this still doesn't quite happen for MacOS, cause it'll easily find the **_/extra_** folder if you got admin powers (which is what you have by default, if the machine is yours), so it'll just prompt you to download there. 

This situation isn't ideal, as I believe the extra folder should not be the best option to install externals. This standard path is the "application specific" folder, and means that it gets lost when you upgrade to a new Pd and throw the old one away. Only advanced users who know what they're doing, and have multiple/different versions of Pd, may want to use it, not all users, who are mostly not savyy enough.

So my suggestion here is to simply suggest in all operating system to first download to the "standard user specific" folder at request. And if the user agrees, it'll create it for them. The biggest change was already done. What I'm proposing simply makes it consistent between cross platforms.

cheers

**Comment by porres** [comment]
Oh, so, there is this recent Pull Request from @danomatika to Pd here: https://github.com/pure-data/pure-data/pull/140 - which may relate to this suggestion a bit.

The thing is that the entry path from that pull request loads the path which is set in deken. But initially that's empty, and only after downloading from deken something shows up there. 

So, it seems that if you initially suggest the user specific folder for download, then that would be there out of the box, instead of an empty field.

**Comment by chr15m** [comment]
@porres sounds good to me.

**Comment by danomatika** [comment]
This should be fixed in #155

**Comment by porres** [comment]
> This should be fixed in #155

If I get this correctly, you mean that it doesn't show an empty filed out of the box anymore, right?

But we'd still need a Pull Request to initially suggest user specific folder for download, right?

**Comment by danomatika** [comment]
>If I get this correctly, you mean that it doesn't show an empty filed out of the box anymore, right?

Yes.

>But we'd still need a Pull Request to initially suggest user specific folder for download, right?

No. When you use deken it asks as it did before which is the (one discussed) preferred default.

**Comment by porres** [comment]
> which is the (one discussed) preferred default.

Not sure what you mean, I don't think the "extra" was ever discussed and agreed as the preferred download, and that's not how deken works. I guess you probably know all this and I'm likely being repetitive, but anyway... deken actually has a priority to download to the user specific folder already, and it'll first prompt you to download it the when first launching Pd, but only if you already have that folder. The second option in this order would be the global folder, and the last is the application specific folder - by the way, this order comes from Pd, not deken, and I think it makes sense. 

What deken does is that if you cannot download it anywhere (that is if you don't have user/global folders or writing privileges for the "extra" folder), it prompts you to download to the user folder. So maybe we could agree now which is the preferred default.

> @porres sounds good to me.

Awesome, and how do you think it'd be best to approach a solution? I see now that when I'm suggesting not to download at first to "extra", this would imply in a change that would prevent deken to first propose that one. 

But if the user wants to download things there, he can just browse there, and deken will remember it between restarts. We now have this another way to set the download path, thanks to Dan, which makes all this much, much easier.

**Comment by danomatika** [comment]
>Not sure what you mean, I don't think the "extra" was ever discussed and agreed as the preferred download, and that's not how deken works.

It's the first "safe" place which is guaranteed to exist and, since it is installed/comes with Pd, is most likely writable, hence it is the default if no deken path as been saved. If you look at the code, it 's basically the first system path that Pd uses.

>I guess you probably know all this and I'm likely being repetitive, but anyway...

Yes.

> deken actually has a priority to download to the user specific folder already, and it'll first prompt you to download it the when first launching Pd, but only if you already have that folder.

This is why the first system path is used as  a reasonable default: it's guaranteed to be there.

>The second option in this order would be the global folder, and the last is the application specific folder - by the way, this order comes from Pd, not deken, and I think it makes sense.

It's different on many platforms so it's more than reasonable to say "here's the default, change if you like." A long part of the discussion on the list, if I remember, was that many people did *not* like the idea of being forced to use a specific folder or even a "standard" folder.

You can argue that all over again but, please, I suggest writing much shorter (& fewer) emails. It's sometimes so much that I honestly don't read very much...

**Comment by porres** [comment]
> A long part of the discussion on the list, if I remember, was that many people did not like the idea of being forced to use a specific folder or even a "standard" folder.
> 
> You can argue that all over again but, please, I suggest writing much shorter (& fewer) emails. 

Well, which discussion you mean? Discussions can tenfold into subtopics and it may seem this was argued before, but this proposal (standard user folder instead of extra) is being first brought up here, so there's no argument to be re edited. However, I've been referring to the "**Deken Install User Experience**" thread on the pd list - which I wasn't involved, so don't blame me if it was long for reading or if it reached nowhere :) 

This thread mentions a previous discussion on automatically creating folders, and the end of it was a proposal to allow the creation back, see: https://lists.puredata.info/pipermail/pd-list/2016-06/115377.html - where I quote @umlaeute "_it seems indeed that the main obstacle for automatic directory creation has gone with this new default user-specific search path. if there is some consensus, i will happily re-enable the directory creation. (but it seems that everybody has gone into sleep mode again, and will only awake after the next release)_"

By the way, the next release would be now ;) 

Yes, I did at one point or another try to revive that discussion on the pd list, to see if it would reach a final conclusion, and it didn't. Yes, if the folder was automatically created, I wouldn't have an issue to point it out here. But note that I'm not reviving any discussion here, I'm not arguing that we should automatically create folders. I'm bringing up a different idea/solution.

cheers

**Comment by danomatika** [comment]
Ok, let me put it another way: "I had no more time for this. I am not going to do anything more on this for now."

I am already chasing bugs on multiple platforms as it is (and wondering why I'm bothering right now...).

**Comment by porres** [comment]
> It's different on many platforms so it's more than reasonable to say "here's the default, change if you like."

yeah, that's what I'm saying, that we should prompt for the user standard path as a default, and people can change that if they like.

**Comment by danomatika** [comment]
How does it not do that already? I really don't understand you now.

As far as I know, the behavior from my Path dialog update PR for a fresh/empty install is:

1. open Pd
2. go to Find Externals
3. search and click on a link
4. a popup appears and asks you if you want to "Install to ../extra?"
5. click No button and choose new path
6. choose new path

<img width="718" alt="screen shot 2017-07-25 at 11 51 57 pm" src="https://user-images.githubusercontent.com/480637/28595469-4d3bcd74-7194-11e7-9e6a-00c86df005fe.png">

With the Path dialog, the behavior is:

1. open Pd
2. go to Preferences -> Path...
3. click on the Browse button
4. choose new path

Are you suggesting a dialog pops up the first time Pd is started? That's pretty annoying, especially to people that might want to keep stuff in the default location as the popup may keep showing up every time they run Pd. I tested adding a popup when you open the Path dialog & the install path is empty, but it's also annoying.

Anyway, the real fix is for deken to choose a default location and/or create that location per platform, not for popups. I am not going to do that right now.

**Comment by porres** [comment]
All I'm suggesting is **which** is the default. Note I said: _"we should prompt for the **user standard path as a default**, and people can change that if they like."_

The issue I opened here also makes it clear, the request is to _"initially suggest **user specific folder** for download"_

Right now it prompts for _extra_ as default. And I did argue in my first comment here how I think that is bad. In short, it's the "application specific" so things get lost when upgrading, probably only advanced users would wanna use it.

Moreover, this is not the folder that gets picked for windows, so it makes things inconsistent in cross platforms. Windows people get prompted for the user specific folder, and that folder is created if it doesn't exist.

I'm repeating myself again, so I'll stop! Hope it is clear now, I can add more details and repeat myself if necessary.

The request in short: Initially prompt for user specific as a default first time around, cause it's better 

thanks for giving attention to this

**Comment by porres** [comment]
> Anyway, the real fix is for deken to choose a default location 

yes, that's what I propose/request, that it chooses **user specific** location

> and/or create that location per platform

It already creates after the user agrees

**Comment by porres** [comment]
> ("extra" is) the first "safe" place which is guaranteed to exist and, since it is installed/comes with Pd, is most likely writable, hence it is the default if no deken path has been saved.

Sorry if I got you wrong, but are you suggesting "extra" is the first default folder for deken? Cause that's not true. It'll prompt for user specific standard path first, if you have it. At least on Mac.

**Comment by Spacechild1** [comment]
this is the situation for the latest pd-0.48.test6 on Windows:

if the user specific folder exist, deken will suggest it, otherwise it will suggest extra. this is bad, as Alex has explained. 
One solution could be that the *installer* automatically creates the user specific folder. 
for the *standalone* distribution (.zip file) this is of course not possible, but also not necessary I think because usually you'd put it to a destination where you have full write access. in this case, installing externals to 'extra' is even an advantage to keep everything local (isn't that the reason why most people would use the .zip file and not the installer?).

**Comment by porres** [comment]
> One solution could be that the installer automatically creates the user specific folder

A solution already in progress is that deken can already create the "user specific folder" (aka user specific "standard path") if it is not there, but it has to suggest it first. But it doesn't suggest when it finds the extra folder as a writable one. Hence, my solution is that it initially suggests the  "user specific folder". It's that simple!

> (isn't that the reason why most people would use the .zip file and not the installer?).

I wouldn't know, but you can always say "no, I don't want that one" and choose another, so it would still work as you say - that is, people who downloaded the zip and want the extra folder, may still do so.

Again, a simple, elegant, and power solution

**Comment by Spacechild1** [comment]
> Hence, my solution is that it initially suggests the "user specific folder".

+1! I think you haven't seen my edit yet :-).

**Comment by porres** [comment]
Wait, I was saying thought in windows it didn't suggest "extra" as it wasn't writable, so instead it'd suggest the "user standard folder" and create it. But now I see it actually prompts you for extra, so it's also bad for windows!!! Maybe something weird is happening to Pd0.48, as it asks to downlaod to Pd48/bin/bin/bin/bin/extra (4 "bin"s? seriously?). 

Well, bottom line is that I thought things were at least ok for windows, but now I see it's also bad for it!

**Comment by porres** [comment]
well, this has been sorted with a new mechanism and many additions from @danomatika to Pd 0.48-0, which is out, so I'm closing this
